kolibri/plugins/hooks.py

```diff
@@ -1,59 +1,326 @@
+
+.. warning::
+    Do not define abstract and registered hooks in the same module. Or to put it
+    in other words: Never put registered hooks in ``<myapp>/hooks.py``. The
+    non-abstract hooks should not be loaded unintentionally in case your
+    application is not loaded but only used to import an abstract definition
+    by an external component!
+
+
+In which order are hooks used/applied?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This is entirely up to the registering class. By default, hooks are applied in
+the same order that the registered hook gets registered! This most likely means
+the order in which ``kolibri_plugin`` is loaded ``=>`` the order in which the
+app is listed in ``INSTALLED_APPS``
+
+
+An example of a plugin using a hook
+-----------------------------------
+
+.. note::
+
+    The example shows a NavigationHook which is simplified for the sake of
+    readability. The actual implementation in Kolibri will defer.
+
+
+Example implementation
+----------------------
+
+Here is an example of how to use a hook in ``myplugin.kolibri_plugin.py``:
+
+
+.. code-block:: python
+
+    from django.db.models
+
+    # This is where the actual abstract hook is defined
+    from kolibri.core.hooks import NavigationHook
+
+    # By inheriting NavigationHook, we tell that we are going to want our
+    # plugin to be part of the hook's activities with the specified attributes.
+    # We only define one navigation item, but defining another one is as simple
+    # as adding another class definition.
+    class MyPluginNavigationItem(NavigationHook):
+
+        label = _("My Plugin")
+        url = reverse_lazy("kolibri:my_plugin:index")
+
+
+And here is the definition of that hook in kolibri.core.hooks:
+
+.. code-block:: python
+
+    from kolibri.plugins.hooks import KolibriHook
+
+
+    class NavigationHook(KolibriHook):
+        \"\"\"
+        Extend this hook to define a new navigation item
+        \"\"\"
+
+        #: A string label for the menu item
+        label = "Untitled"
+
+        #: A string or lazy proxy for the url
+        url = "/"
+
+        @classmethod
+        def get_menu(cls):
+            menu = {}
+            for inheritor in cls.inheritors:
+                menu[inheritor.label] = url
+            return menu
+
+        class Meta:
+
+            abstract = True
+
+
+Usage of the hook
+-----------------
+
+Inside our templates, we load a template tag from navigation_tags, and this
+template tag definition looks like this:
+
+.. code-block:: python
+
+    from kolibri.core.hooks import NavigationHook
+
+    @register.assignment_tag()
+    def kolibri_main_navigation():
+
+        for label, url in NavigationHook.get_menu().items():
+            yield label, url
+
+
+.. code-block:: html
+
+    {% load kolibri_tags %}
+
+    <ul>
+    {% for navigation in kolibri_main_navigation %}
+        <li><a href="{{ navigation.menu_url }}">{{ navigation.menu_name }}</a></li>
+    {% endfor %}
```

Comment:
a few comments on this example:

* I'm a little confused by how it works. The tag has `yield label, url` but names used in the template are `menu_name` and `menu_url`. Where do these names come from?
* Looks like `assignment_tag` is are [deprecated](https://docs.djangoproject.com/en/1.9/howto/custom-template-tags/#assignment-tags)
* I think one of the most common use-cases will be bootstrapping data into the template, rather than rendering HTML tags (which I'm hoping will be done almost entirely by javascript.) [Here's one example](https://github.com/indirectlylit/kolibri/blob/vue/kolibri/core/templatetags/kolibri_tags.py#L28)